build_tools/third_party/cuda/CMakeLists.txt

```diff
@@ -6,16 +6,67 @@
 
 set(CUDA_HEADERS_API_ROOT "${IREE_ROOT_DIR}/third_party/cuda/include")
 
-external_cc_library(
-  PACKAGE
-    cuda_headers
-  NAME
-    cuda_headers
-  ROOT
-    ${CUDA_HEADERS_API_ROOT}
-  HDRS
-    "cuda.h"
-  INCLUDES
-    ${CUDA_HEADERS_API_ROOT}
-)
+unset(CUDA_INCLUDE_DIR CACHE)
+unset(CUPTI_INCLUDE_DIR CACHE)
 
+# Check for the CUDA install and enable if possible. Pull relevant headers into the build
+if(IREE_HAL_DRIVER_CUDA AND IREE_ENABLE_RUNTIME_TRACING)
+  include(CheckLanguage)
+  check_language(CUDA)
+  if (CMAKE_CUDA_COMPILER)
+    enable_language(CUDA)
+  else()
+    message(SEND_ERROR
+      "Building IREE with the CUDA driver requires a working CUDA install.")
```

Comment:
channeling Ben's earlier comment from the C source:
> we don't want to require cupti - if the library fails to load you need to continue without it

I don't think this should be a blocking error in the CMake configuration.

you could add a C/C++ define like `IREE_HAVE_CUPTI` based on the existence of the `CMAKE_CUDA_COMPILER` (or some other) variable then use that to switch in the source files 


I'm already hitting this annoying warning:
```
[cmake] CMake Warning at CMakeLists.txt:285 (message):
[cmake]   IREE's regression test suite requires PyYAML to run all tests.  It is not
[cmake]   installed, so some tests will be disabled.
```

and this PR adds a new error:
```
[cmake] -- Looking for a CUDA compiler
[cmake] -- Looking for a CUDA compiler - NOTFOUND
[cmake] CMake Error at build_tools/third_party/cuda/CMakeLists.txt:19 (message):
[cmake]   Building IREE with the CUDA driver requires a working CUDA install.
[cmake] 
[cmake] 
[cmake] -- Unable to find cuda_runtime.h in "D:/dev/projects/iree/build_tools/third_party/cuda/undefined/include" for CUDAToolkit_INCLUDE_DIR.
[cmake] -- Could NOT find CUDAToolkit (missing: CUDAToolkit_INCLUDE_DIR) (found version "11.4.120")
```

(I think I have CUDA installed on this Windows machine and can help CMake find it, but this code should still be tolerant to the library not being found)

Reply:
Actually, I need to update that error message, It should indicate that you need a working CUDA install for the _CUDA driver with tracing_ enabled (that code path is guarded on those flags), not just CUDA alone. 

Right now, if the CUPTI library isn't found, the rest of the CUPTI code is enabled (based on those two flags) and will fail to compile anyway. 

I could define a global flag (which would require I do more CMake.... T_T) that checks early on if you have CUDA with CUPTI installed based on enabling the CUDA backend + tracing and then guard on that. If you don't meet the requirements, you simply don't get CUDA tracing. 

If I go this rout, can I define the flag in the base CMakeLists.txt?

Reply:
It might make sense to flip `IREE_HAL_DRIVER_CUDA` off if if the CUDA SDK is not configured then. I _did_ install CUDA on this machine through the official installer with default paths... but those CMake checks failed for me :P. I see some people suggesting setting `CMAKE_CUDA_COMPILER` explicitly, but maybe there's a way to get this working on Windows without that.

Reply:
I'm trying this on Windows now.

* For `find_package(CUDAToolkit)`, I needed to update CMake beyond 3.20.3 (I was on 3.20.0) to get past https://gitlab.kitware.com/cmake/cmake/-/issues/22227 in VSCode
* `enable_language(CUDA)` is still failing on my machine. https://stackoverflow.com/questions/56636714/cuda-compile-problems-on-windows-cmake-error-no-cuda-toolset-found suggests that the CUDA installer does strange things with MSVC and a reinstall (since I've updated my MSVC a few times) may help...

I'll reiterate that this (the CUDA driver, the CUDA tracing support, etc.) should all be optional. There's just too much configuration pain on each platform to enforce it by default.

Reply:
+1 on optional cuda

Reply:
Reinstalling the CUDA SDK helped `enable_language(CUDA)` succeed in my case. I was able to build `iree-run-module` with CUDA and tracing enabled with just the typo fix below (`CUDA_INCLUD_DIR` -> `CUDA_INCLUDE_DIR`) and some tweaks to the `#if` / tracing macros mentioned in another comment.

I'd like to use this style here: 
https://github.com/google/iree/blob/3d9731439a09f97748dc8ff1bd852dc21550dfdd/iree/testing/vulkan/CMakeLists.txt#L22-L26

* try finding, possibly with `QUIET`
* report issues at `VERBOSE` level so people can set `-DCMAKE_MESSAGE_LOG_LEVEL=STATUS` to skip over the output when not debugging configuration issues